File: 76.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if t == "":
            return ""
        m = {}
        tm = {}
        matches = 0
        ans = [-1, -1]
        lenAns = float("infinity")
        for c in t:
            tm[c] = 1 + tm.get(c, 0)
        l = 0
        for r, curr in enumerate(s):
            m[curr] = 1 + m.get(curr, 0)
            if curr in tm and m[curr] == tm[curr]:
                matches += 1
            while matches == len(tm):
                if r - l + 1 < lenAns:
                    ans = [l, r]
                    lenAns = r - l + 1
                m[s[l]] -= 1
                if s[l] in tm and m[s[l]] < tm[s[l]]:
                    matches -= 1
                l += 1
        l, r = ans
        return s[l : r + 1] if lenAns != float("infinity") else ""


print(Solution().minWindow("ADOBECODEBANC", "ABC"))
# print(Solution().minWindow("aaaaaaaaaaaabbbbbcdd", "abcdd"))

File: 76.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
  string minWindow(string s, string t) {
    if (t.size() == 0) {
      return "";
    }
    unordered_map<char, int> tm;
    unordered_map<char, int> m;
    for (auto c : t) {
      ++tm[c];
    }
    int l = 0, lenAns = 0xfffffff, matches = 0;
    pair<int, int> ans = {-1, -1};
    for (int r = 0; r < s.size(); ++r) {
      char curr = s[r];
      ++m[curr];
      if (tm.find(curr) != tm.end() && tm[curr] == m[curr]) {
        ++matches;
      }
      while (matches == tm.size()) {
        if (r - l + 1 < lenAns) {
          ans = {l, r};
          lenAns = r - l + 1;
        }
        --m[s[l]];
        if (tm.find(s[l]) != tm.end() && m[s[l]] < tm[s[l]]) {
          matches -= 1;
        }
        ++l;
      }
    }
    if (lenAns == 0xfffffff) {
      return "";
    }
    string answer;
    for (auto it = s.begin() + ans.first; it < s.begin() + ans.second + 1;
         ++it) {
      answer.push_back(*it);
    }
    return answer;
  }
};

int main() {
  auto a = Solution().minWindow("ADOBECODEBANC", "ABC");
  cout << a << endl;
}